Release: migrate release process to release-it

timmywil opened the pull request:
### Summary ###

- This migrates the release process from jquery-release to [release-it](https://github.com/release-it/release-it) and some custom scripts.
- Also, now that we no longer need to build on Node 10 or maintain jquery-release scripts, this PR migrates all files in the build folder to use ESM.
- Moved the dist and dist-module wrapper files to `wrappers/` folders to make packaging and cleaning release-only files easier. All references to those files have been updated and it should have no effect on the exports.

*Authors*
- Checking and updating authors has been migrated
  to a custom script in the repo

*Changelog*
- changelogplease is no longer maintained; replace with custom script
- generate changelog in markdown for GitHub releases
- generate changelog in HTML for blog posts

*dist*
- clone dist repo, copy files, and commit/push
- commit tag with dist files on main branch;
  remove dist files from main branch after release

*cdn*
- clone cdn repo, copy files, and commit/push
- create versioned and unversioned copies in cdn/
- generate md5 sums and archives for Google and MSFT


### Checklist ###
<!--
Mark an `[x]` for completed items, if you're not sure leave them unchecked and we can assist.
-->

* [x] New tests have been added to show the fix or feature works
* [x] If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com

<!--
Thanks! Bots and humans will be around shortly to check it out.
-->

This PR currently has feature parity with jquery-release, but these are a few things I'd like to add:

### Future Improvements

- [x] Add workflow to verify new release builds are reproducible
- [x] Document updated release process in `build/release/README.md`
- [x] Generate contributors list with links to GH profiles
- [ ] (stable only) Update version in jquery.com repo
- [ ] (stable only) Update version in jquery-wp-content
- [ ] (stable only) Update version in api.jquery.com demos (entries2html.xsl)
- [ ] (optional) Close GH milestone

timmywil commented:
I've tested every step of this. The final test will be to try a real release, but we can wait until after the next beta release.

timmywil commented:
I realized we can go ahead and review this. I'd like to add a couple things that jquery-release didn't have for stable releases, but those things are optional and could be added in future PRs.

mgol commented on build/release/dist.js:
```diff
@@ -1,190 +1,124 @@
-		"src",
-		"LICENSE.txt",
-		"AUTHORS.txt",
-		"dist/package.json",
```
How are we generating `package.json` now?

timmywil replied:
This is referring to the `package.json` in the `dist` folder, which sets the module type. Everything in the `dist` folder is included, so it's no longer being specified explicitly.

mgol commented on build/release/post-release.sh:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env sh
```
Let's add some flags here, like `-u`, `-e`, `-o pipefail`?

mgol replied:
This seems like overkill for `sh`, why not:
```suggestion
#!/bin/sh
```
?

mgol reviewed (changes requested):
It'll be great to have reproducible builds!

I added some comments.

timmywil commented on .github/workflows/verify-release.yml:
```diff
@@ -0,0 +1,33 @@
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Version to verify (>= 4.0.0-beta.1)'
```
That's not normally how they are numbered. The first beta should have been 4.0.0-beta.0.

mgol replied:
What's the source for this naming strategy?

timmywil replied:
Official sources allow for more wiggle room in pre release versions. However, it's moot because this is how release-it works by default and it's not really changeable.

timmywil replied:
As discussed in the meeting, will update so the next release is `.2`. This may require manual edits to the package.json before initial pre-releases, but we can do that.

timmywil commented on .gitignore:
```diff
@@ -13,26 +12,27 @@ tmp
 /test/data/qunit-fixture.js
 
-# Ignore BrowserStack files
+# Release artifacts
+changelog.*
```
Everything but changelogs go into the `tmp/` folder. Archives will be under `tmp/archives`. I had almost everything top-level originally, but that was a lot of `.gitignore` entries so I put everything under `tmp/` instead. The changelogs seemed a little different because `changelog.md` is copied to jquery-dist top-level and changelog.html is only for the blog post.

mgol replied:
Why not put `changelog.html` in `tmp` then?

timmywil replied:
I'm not sure that's more helpful. You'd still need the extra `.gitignore` entry and `.*` is even shorter than `.md`.

timmywil replied:
That said, I could just stick both in `tmp`.

timmywil replied:
Actually, could we leave it? I was thinking we could commit the changelog.md to the tag as well, so it would need to stay top-level. And if the `.md` is top-level, it's a little more convenient for me to have the `.html` top-level. I can foresee myself continuing the post release, which removes everything before I've copied the HTML, but that's much less likely if I see the HTML top-level. Yes, I can always generate it again, but I like that it's more in my face.

mgol replied:
We discussed it on the meeting; we're leaving it as-is.

timmywil commented on .release-it.cjs:
```diff
@@ -0,0 +1,38 @@
+"use strict";
```
Yes, I noted in the PR that I planned on updating the release docs after this PR is merged. I figured I'd update the wiki, but it could be a README in the `build/release`?

mgol replied:
A README in `build/release` works for me.

timmywil commented on build/release/verify.js:
```diff
@@ -0,0 +1,199 @@
+
+			// Get file contents from tarball
+			const { stdout: npmContents } = await exec(
+				`tar -xOf ${ npmTarballPath } package/${ file.path }/${ file.name }`
```
no, this works on windows as well

timmywil commented on .release-it.cjs:
```diff
@@ -0,0 +1,38 @@
+	git: {
+		changelog: "npm run release:changelog -- ${from} ${to}",
+		commitMessage: "Release: ${version}",
+		getLatestTagFromAllRefs: true,
```
Hmm, I didn't think it would work at all without this, but maybe you're right.

timmywil replied:
Will remove.

timmywil replied:
Wait, I think it's needed the first time though because 4.0.0-beta is detached.

timmywil replied:
We can remove immediately after.

timmywil commented on .release-it.cjs:
```diff
@@ -0,0 +1,38 @@
+			"sed -i 's/main/AUTHORS.txt/${version}/AUTHORS.txt/' package.json",
+		"after:release": `./build/release/post-release.sh \${version} ${ blogURL }`
+	},
+	git: {
```
Yes, that was one of my goals.

timmywil commented on build/command.js:
```diff
@@ -1,8 +1,6 @@
-const { build } = require( "./tasks/build" );
-const yargs = require( "yargs/yargs" );
-const slimExclude = require( "./tasks/lib/slim-exclude" );
+import yargs from "yargs/yargs";
```
I also considered that, but `yargs` does provide us more than args parsing. It also gives us a pretty help menu.

timmywil commented on build/release/archive.js:
```diff
@@ -0,0 +1,59 @@
+		return sum.join( "\n" );
+	}
+
+	const { stdout } = await exec( "md5 -r " + files.join( " " ), { cwd: folder } );
```
I'm not sure, but let's look into that when we need to support Linux.

mgol replied:
It will probably be needed when we automate the release via GitHub Actions.

For build breaking Linux would not be acceptable but I guess for a release it is.

timmywil commented on build/release/cdn.js:
```diff
@@ -1,175 +1,127 @@
+				.replace( "jquery", "jquery-" + version );
+			const minFilename = filename.replace( ".js", ".min.js" );
+
+			await exec( `cp -f ${ file } ${ versionedFolder }/${ filename }` );
```
Yes

timmywil commented on build/release/cdn.js:
```diff
@@ -1,175 +1,127 @@
 
-			archiver.pipe( output );
+async function cdn() {
+	await clean();
```
This was useful when testing this step on its own. It should be a noop for real releases, but I'd like to leave it.

timmywil commented on build/release/changelog.js:
```diff
@@ -0,0 +1,163 @@
+const rendsWithHash = /#$/;
+const rcherry = / \(cherry picked from commit [^)]+\)/;
+const rcommit = /Fix(?:e[sd])? ((?:[a-zA-Z0-9_-]{1,39}\/[a-zA-Z0-9_-]{1,100}#)|#|gh-)(\d+)/g;
+const rcomponent = /^([^ :]+):\s*([^\n]+)/;
```
I actually wasn't aware we did that. I've never done that. Perhaps it'd make things simpler to stick to one component per commit.

timmywil replied:
You know, what happens already isn't bad:

```
## CSS

- Tests: Fix tests & support tests under CSS Zoom
```

timmywil commented on build/release/changelog.js:
```diff
@@ -0,0 +1,163 @@
+function getTicketsForCommit( commit ) {
+	var tickets = [];
+
+	commit.replace(
```
Is it weird? I thought it was convenient.

mgol replied:
It's like using `Array#map` without using its output when you really want `Array#forEach`. It obscures intent - `String#replace` is usually stateless, just like `Array#map`.

Why not use `String#matchAll` followed by `Array#map` with your logic?

timmywil replied:
It feels weirder to me to do that because it's looping twice instead of once. Granted, a micro-optimization that doesn't matter for our sample size, but it's not the way I've trained my brain. 😛 I guess this is one of those cases where the ideal doesn't really exist. And yet, if there were some method that was more like `forEach`, the implementation of it might be about identical to `replace`.

timmywil commented on build/release/changelog.js:
```diff
@@ -0,0 +1,163 @@
+	return a.localeCompare( b );
+}
+
+/**
+ * Remove all revert commits and the commit it is reverting
+ */
```
I'm not sure I follow. Does it not already do that? I think the final changelog would compare to the latest RC. But, you bring up a good point that we'll probably want a full 4.0 changelog for the blog post, which we can prepare manually by using `npm run release:changelog -- 3.7.1 4.0.0`.

timmywil commented on build/release/changelog.js:
```diff
@@ -0,0 +1,163 @@
+	const bComponent = rcomponent.exec( b );
+
+	if ( aComponent && bComponent ) {
+		return aComponent[ 1 ].localeCompare( bComponent[ 1 ] );
```
I liked it because of how terse it was, but I suppose this could directly affect us if you have another language as your default, like Polish.

timmywil commented on build/release/changelog.js:
```diff
@@ -0,0 +1,163 @@
+const rbeforeHash = /.#$/;
+const rendsWithHash = /#$/;
+const rcherry = / \(cherry picked from commit [^)]+\)/;
+const rcommit = /Fix(?:e[sd])? ((?:[a-zA-Z0-9_-]{1,39}\/[a-zA-Z0-9_-]{1,100}#)|#|gh-)(\d+)/g;
```
This regex was copied directly from changelogplease, so it's what we had before. I'd say let's keep it the same for now and look into expanding it in a future PR.

timmywil commented on build/release/changelog.js:
```diff
@@ -0,0 +1,163 @@
+
+		const match = /\*\s*Revert "([^"]*)"/.exec( commit );
+
+		// Ignore double reverts
```
Heh, I know. I think that has happened once, but I edited it manually.

timmywil commented on build/release/post-release.sh:
```diff
@@ -0,0 +1,51 @@
+
+# Remove release files from tracking
+npm run release:clean
+git commit -m "Release: remove release-only files from main branch"
```
That's part of the regular release process in the [`git` plugin](https://github.com/release-it/release-it/blob/main/docs/git.md)

timmywil commented on .release-it.cjs:
```diff
@@ -0,0 +1,38 @@
+		tokenRef: "JQUERY_GITHUB_TOKEN"
+	},
+	npm: {
+		publish: true
```
Those are already the defaults, but they can be modified using command line args.

timmywil commented on build/tasks/build.js:
```diff
@@ -38,14 +37,14 @@ const removeWith = {
 
 async function read( filename ) {
-	return fs.readFile( path.join( srcFolder, filename ), "utf8" );
+	return fs.readFile( path.join( "./src", filename ), "utf8" );
```
Yes, it works from any directory as long as you're using npm scripts.

timmywil commented on build/release/post-release.sh:
```diff
@@ -0,0 +1,51 @@
+fi
+
+# Push files to cdn repo
+npm run release:cdn $1
```
I suppose we could, but with npm scripts as the entry point, the working directory is always the same.

timmywil commented on build/release/verify.js:
```diff
@@ -0,0 +1,199 @@
+	if ( rstable.test( version ) ) {
+		await Promise.all(
+			release.files.map( async( file ) => {
+				console.log( file.name );
```
Nice catch.

timmywil commented on build/release/verify.js:
```diff
@@ -0,0 +1,199 @@
+
+	// Pack the npm tarball
+	console.log( `Packing jQuery ${ version }...` );
+	await exec( "npm pack", { cwd: releaseFolder } ).then( ( { stdout } ) =>
+		console.log( stdout )
+	);
```
I used `.then` to avoid renaming, but I'll just rename.

timmywil commented on build/release/verify.js:
```diff
@@ -0,0 +1,199 @@
+
+	// Pack the npm tarball
+	console.log( `Packing jQuery ${ version }...` );
+	await exec( "npm pack", { cwd: releaseFolder } ).then( ( { stdout } ) =>
```
To compare checksums of the tarballs.

mgol commented on build/release/README.md:
```diff
@@ -0,0 +1,63 @@
+
+Stable releases have a few more steps:
+
+1. Close the milestone matching the current release: https://github.com/jquery/jquery/milestones.
```
...and create one for the next patch release if one already doesn't exist? I guess it will most of the time...

timmywil replied:
I think that's covered in the first step above:

> Ensure all milestoned issues/PRs are closed, or reassign to a new milestone.

mgol replied:
There's an edge case possible where we just fix everything for a release and avoid creating a new one and then there's no milestone ready for new tickets. It's easy to create one when needed, though, so maybe it's good as-is.

timmywil replied:
I see

mgol commented on build/release/README.md:
```diff
@@ -0,0 +1,69 @@
+2. Update jQuery on https://github.com/jquery/jquery-wp-content.
+3. Update jQuery on https://github.com/jquery/blog.jquery.com-theme.
+4. Update latest jQuery version for [healthyweb.org](https://github.com/jquery/healthyweb.org/blob/main/wrangler.toml).
+5. Update the shipping version on [jquery.com home page](https://github.com/jquery/jquery.com).
```
I’d use `1.` for all points; the diffs are much better when you need to insert something in the middle. Some standards repos are using this strategy, probably for the same reason.

timmywil replied:
I would normally do that, but with code examples between the numbered list items, the numbers restart. It would work for the first 5 here, but not for any of the below items.

mgol replied:
That just means they aren't formatted properly - so it's even better to `1`-index as then you uncover those bugs immediately. You should indent the code snippets (the triple backticks should also be indented) and they will be pulled into the list.

This should be done as even though the numbers may look correct, we have multiple `<ol>` elements so semantics of the lists are incorrect.

mgol replied:
The same applies to the CDN table at the bottom - since it refers to the 7th point on the list, it should exist within that list.

timmywil replied:
But indenting the code block ruins the syntax highlighting, at least in vs code preview. I haven't pushed it to github to check there.

mgol replied:
I think semantics are more important - as they affect accessibility - than previews in VS Code. I've used indented code blocks in lists on GitHub and it works perfectly. Maybe the indentation was still not right? I can check later if you don't manage to make it work.

No need to push it to GitHub to test, BTW. You can just put it in a comment, switch to the Preview tab and inspect the HTML.

timmywil replied:
Take that back. Indenting code blocks or unordered list elements only ruins the markdown code syntax highlighting, and I must have assumed the preview was broken, but it seems to work. However, there are downsides. It reduces the max line length fairly significantly for code blocks.

timmywil replied:
Seems to be a fair trade-off. I even misnumbered the last item, so this fixes it.

timmywil replied:
GitHub has condensed any unordered list items and put them on one line. I'm seeing if we can fix that with new lines.

timmywil replied:
I don't know if it's actually an accessibility issue, though. The multiple `<ol>` elements were still readable by screen readers. Subsequent ones had a different `start` attribute so the numbers lined up.

mgol reviewed (approved):
LGTM. Remaining comments can be addressed separately later.

mgol commented on build/release/README.md:
```diff
@@ -0,0 +1,69 @@
+2. Verify all tests are passing in [CI](https://github.com/jquery/jquery/actions).
+3. Run any release-only tests, such as those in the [`test/integration`](../../test/integration/) folder.
+4. Ensure AUTHORS.txt file is up to date (this will be verified by the release script).
+  - Use `npm run authors:update` to update.
```
This should also be indented more. Right now, it creates a separate `<ul>` at the same level as the `<ol>` above.

I'd recommend you to look at the rendered HTML and see if the structure makes sense by yourself. You'll uncover all these semantic bugs pretty quickly then.

timmywil replied:
The extra indent always looked like it registered in code as plain text instead of a list item, but it seems to work.

timmywil replied:
The extra indent doesn't work on GitHub.

timmywil replied:
Adding a newline but removing the extra indent seems to work though.

timmywil commented:
Changes since last approval:

1. Now generating a `contributors.html` that includes author names and links to their GH profiles for use in the blog post "Thanks" section. So much better than what I was using before!
2. Added a "Setup" section to the release README, which includes basic setup instructions like cloning and installing dependencies, but more importantly setup instructions for logging into npm and ensuring a github token is set in the environment, with an easy link to create a new token if necessary.
3. Updated release-it to 17.5.0, which includes the new `--preReleaseBase` option, which should be set to `1` for the first pre-release of a kind (e.g. `--preRelease=beta --preReleaseBase=1` to release `.beta.1` instead of `.beta.0`). This has also been documented.
4. Updated AUTHORS.txt to prepare for the upcoming release.

mgol commented on build/release/README.md:
````diff
@@ -0,0 +1,116 @@
+	git push origin main --tags
+	```
+
+6. Update the version used in [jQuery docs demos](https://github.com/jquery/api.jquery.com/blob/main/entries2html.xsl).
+
+7. Email archives to CDNs.
````
These two should also be `1.`s.

mgol reviewed (approved):
Still LGTM after the latest changes. Just a small note on leftover larger numbers in lists.
